server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java

```diff
@@ -0,0 +1,167 @@
+import java.util.Map;
+import java.util.Set;
+
+import static org.elasticsearch.cluster.node.DiscoveryNodeFilters.OpType.AND;
+import static org.elasticsearch.cluster.node.DiscoveryNodeFilters.OpType.OR;
+import static org.elasticsearch.cluster.node.DiscoveryNodeRole.INDEX_ROLE;
+import static org.elasticsearch.cluster.node.DiscoveryNodeRole.SEARCH_ROLE;
+import static org.elasticsearch.cluster.routing.allocation.decider.FilterAllocationDecider.CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.decider.FilterAllocationDecider.CLUSTER_ROUTING_INCLUDE_GROUP_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.decider.FilterAllocationDecider.CLUSTER_ROUTING_REQUIRE_GROUP_SETTING;
+
+/**
+ * For an index of n shards hosted by a cluster of m nodes, a node should not host
+ * significantly more than n / m shards. This allocation decider enforces this principle.
+ * This allocation decider excludes any nodes flagged for shutdown from consideration
+ * when computing optimal shard distributions.
+ */
+public class IndexBalanceAllocationDecider extends AllocationDecider {
+
+    private static final Logger logger = LogManager.getLogger(IndexBalanceAllocationDecider.class);
+    private static final String EMPTY = "";
+
+    public static final String NAME = "index_balance";
+
+    private final IndexBalanceConstraintSettings indexBalanceConstraintSettings;
+    private final boolean isStateless;
+
+    private volatile DiscoveryNodeFilters clusterRequireFilters;
+    private volatile DiscoveryNodeFilters clusterIncludeFilters;
+    private volatile DiscoveryNodeFilters clusterExcludeFilters;
+
+    public IndexBalanceAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.indexBalanceConstraintSettings = new IndexBalanceConstraintSettings(clusterSettings);
+        setClusterRequireFilters(CLUSTER_ROUTING_REQUIRE_GROUP_SETTING.getAsMap(settings));
+        setClusterExcludeFilters(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING.getAsMap(settings));
+        setClusterIncludeFilters(CLUSTER_ROUTING_INCLUDE_GROUP_SETTING.getAsMap(settings));
+        clusterSettings.addAffixMapUpdateConsumer(CLUSTER_ROUTING_REQUIRE_GROUP_SETTING, this::setClusterRequireFilters, (a, b) -> {});
+        clusterSettings.addAffixMapUpdateConsumer(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING, this::setClusterExcludeFilters, (a, b) -> {});
+        clusterSettings.addAffixMapUpdateConsumer(CLUSTER_ROUTING_INCLUDE_GROUP_SETTING, this::setClusterIncludeFilters, (a, b) -> {});
+        isStateless = DiscoveryNode.isStateless(settings);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (indexBalanceConstraintSettings.isDeciderEnabled() == false || isStateless == false || hasNoFilters() == false) {
+            return Decision.single(Decision.Type.YES, NAME, "Decider is disabled.");
+        }
```

Comment:
If it's disabled for stateful, I wonder if we could just configure it to be added in the stateless plugin? Then we wouldn't need to check `isStateless` every time.

Reply:
Also `hasNoFilters() == false` is harder to read than `filtersAreConfigured()` or similar. It's Friday here, I can't process a double-negative.